### PR TITLE
Add env for jwks refresh interval configuration 

### LINKED ISSUE #29408: Option to configure JWKs public key refresh interval
**Describe the feature request**

Reviving this one that got closed by a bot because nobody attended to it:

https://github.com/istio/istio/issues/23409

### LINKED ISSUE #29436: jwks cache related errors should result in refetch of jwks_uri
**Describe the feature request**

It would be nice if istio would refetch jwks data upon jwks cache related errors, as it might have stale data.
( We now get `Jwks doesn't have key to match kid or alg from Jwt` errors after we redeploy keykloak with a fresh db)

NOTE: There seems to be a hardcoded 20 minutes TTL on the jwks gotten from the jwks_uri (see #27097)

In the mean time it would be nice to be pointed to a workaround as to how to kill that cache manually. It seems to extend it's ttl upon requesting the invalid route, making it even worse. (It seems one has to wait exactly 20 minutes while hoping no request hits that route, in order for the 20 minutes to fully pass.)

[x] Docs
[ ] Installation
[x] Networking
[ ] Performance and Scalability
[ ] Extensions and Telemetry
[x] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

**Additional context**

### PR DESCRIPTION
**Pull request by xulingqing**
[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[X] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

Pull Request Attributes

Please check any characteristics that apply to this pull request. 

[X] Does not have any changes that may affect Istio users.

### CODE REVIEW

**Review by howardjohn — COMMENTED** [review]
Is this the right API? I found 2 different usages in the wild:

https://github.com/coreos/go-oidc/blob/v2/jwks.go#L223 uses Cache-control header from jwks server

https://openid.net/specs/openid-connect-core-1_0.html#RotateSigKeys suggests rotating when encountering a unknown `kid` value - since we are not verifying the jwts we probably cannot do this

**Comment by howardjohn** [comment]
also the cache-control part was removed in v3 of that library: https://github.com/coreos/go-oidc/pull/259

**Comment by xulingqing** [comment]
> Is this the right API? I found 2 different usages in the wild:
> 
> https://github.com/coreos/go-oidc/blob/v2/jwks.go#L223 uses Cache-control header from jwks server
> 
> https://openid.net/specs/openid-connect-core-1_0.html#RotateSigKeys suggests rotating when encountering a unknown `kid` value - since we are not verifying the jwts we probably cannot do this

This PR focus on make the JwksPubKey refresh configurable. It's different from the Cache-control header thing. The feature request is from issue#29408.

**Comment by howardjohn** [comment]
> This PR focus on make the JwksPubKey refresh configurable. 

Understood. My concern is may its not something that should be configurable, and maybe polling every 20 minutes is also not correct. We should be looking to see what the standards or well known implementations are doing.

**Comment by howardjohn** [comment]
cc @yangminzhu @costinm probably know more than me about ^

**Comment by yangminzhu** [comment]
> > This PR focus on make the JwksPubKey refresh configurable.
> 
> Understood. My concern is may its not something that should be configurable, and maybe polling every 20 minutes is also not correct. We should be looking to see what the standards or well known implementations are doing.

I'm not aware of any standard around this. Different provider may have different key rotation interval especially if the user owns the jwks key provider and configures the rotation interval themselves.

Some well-known provider use the cache-control but even so it's still a bit risky to use it in any way because some providers may just rotate the key early than the `max-age` in the cache-control and it could change in every response which makes it much more complicated in the implementation, See https://discuss.dgraph.io/t/authentication-using-jwkurl/10455/9

So I think it's worth to make this at least configurable so that the user can fine-tune to better fit their jwks provider.

File: pilot/pkg/features/pilot.go [hunk 1]
```diff
@@ -406,4 +406,10 @@ var (
 	PilotEnableLoopBlockers = env.RegisterBoolVar("PILOT_ENABLE_LOOP_BLOCKER", true,
 		"If enabled, Envoy will be configured to prevent traffic directly the the inbound/outbound "+
 			"ports (15001/15006). This prevents traffic loops. This option will be removed, and considered always enabled, in 1.9.").Get()
+
+	PilotJwtPubKeyRefreshInterval = env.RegisterDurationVar(
+		"PILOT_JWTPUBKEY_REFRESH_INTERVAL0",
```
Comment (yangminzhu) [inline pilot/pkg/features/pilot.go:411]: The name should be `PILOT_JWT_PUB_KEY_REFRESH_INTERVAL`

File: pilot/pkg/features/pilot.go [hunk 2]
```diff
@@ -406,4 +406,10 @@ var (
...
+		20*time.Minute,
+		"The refresh interval for istiod to fetch and cache the jwks public key. ",
```
Comment (yangminzhu) [inline pilot/pkg/features/pilot.go:413]: "The interval for istiod to fetch the jwks_uri for the jwks public key.`

**Review by yangminzhu — COMMENTED** [review]
Could you also add a release note including a small example of how to set the added environment variable? thanks

File: pilot/pkg/features/pilot.go [hunk 3]
```diff
@@ -406,4 +406,10 @@ var (
 	PilotEnableLoopBlockers = env.RegisterBoolVar("PILOT_ENABLE_LOOP_BLOCKER", true,
 		"If enabled, Envoy will be configured to prevent traffic directly the the inbound/outbound "+
 			"ports (15001/15006). This prevents traffic loops. This option will be removed, and considered always enabled, in 1.9.").Get()
+
+	PilotJwtPubKeyRefreshInterval = env.RegisterDurationVar(
+		"PILOT_JWTPUBKEY_REFRESH_INTERVAL0",
+		20*time.Minute,
```
Comment (costinm) [inline pilot/pkg/features/pilot.go:412]: Seems far too low as a default. 1 week maybe - do we have data on how often typical OIDC are refreshing their root ?

**Review by costinm — COMMENTED** [review]
Looks good except the comments ( name and default )

File: pilot/pkg/features/pilot.go (hunk 3 above)
Comment (yangminzhu) [reply pilot/pkg/features/pilot.go:415]: That's not changed in the PR, there is no common standard around the default, some provider (e.g. Firebase) has a expiration interval of several hours.

File: releasenotes/notes/30014.yaml [hunk 4]
```diff
@@ -0,0 +1,10 @@
+apiVersion: release-notes/v2
+kind: feature
+area: security
+
+releaseNotes:
+- |
+  **Added** Add an environment variable PILOT_JWT_PUB_KEY_REFRESH_INTERVAL for config the interval of
+  istiod to fetch the jwks_uri for the jwks public key. User can use set the refresh interval with command
+  `--set values.pilot.env.PILOT_JWT_PUB_KEY_REFRESH_INTERVAL=20m` or `=10s` while istioctl installation.
```
Comment (yangminzhu) [inline releasenotes/notes/30014.yaml:9]: 1. Remove the extra `Add`
2. User can set the refresh interval with command `--set values.pilot.env.PILOT_JWT_PUB_KEY_REFRESH_INTERVAL=<duration>`, the default interval is `20m`.

**Review by yangminzhu — APPROVED** [review]

**Comment by yangminzhu** [comment]
any updates for this PR?

**Review by howardjohn — APPROVED** [review]

**Review by morvencao — APPROVED** [review]

**Pull request merged**